test/unit/StakingProducts/recalculateEffectiveWeight.js

```diff
@@ -0,0 +1,447 @@
+
+    // buy all cover on all products at 50% target weight (1/2 max)
+    const allocationPromises = [];
+    for (const productId of DEFAULT_PRODUCTS) {
+      allocationPromises.push(allocateCapacity.call(this, { coverBuyer, amount: coverBuyAmount, productId }));
+    }
+    await Promise.all(allocationPromises);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(
+      await stakingProducts.getTotalTargetWeight(this.poolId),
+    );
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(1000);
+
+    // burn half of active stake: leaving 1/2 of the capacity, so effective weight should now be maxed out
+    const activeStake = await stakingPool.getActiveStake();
+    await burnStake.call(this, { amount: activeStake.div(2), start });
+
+    // recalculate effective weight
+    await stakingProducts.recalculateEffectiveWeights(this.poolId, DEFAULT_PRODUCTS);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(2000);
+    expect(await stakingProducts.getTotalTargetWeight(this.poolId)).to.be.equal(1000);
+
+    // lowering target weight shouldn't change effective weight
+    await setStakedProducts.call(this, { productIds: DEFAULT_PRODUCTS, targetWeight: 1 });
+    await stakingProducts.recalculateEffectiveWeights(this.poolId, DEFAULT_PRODUCTS);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(2000);
+    expect(await stakingProducts.getTotalTargetWeight(this.poolId)).to.be.equal(20);
+
+    // raising target weight also shouldn't change effective weight
+    await setStakedProducts.call(this, { productIds: DEFAULT_PRODUCTS, targetWeight: 100 });
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(
+      await stakingProducts.getTotalTargetWeight(this.poolId),
+    );
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(2000);
+
+    // burn half the remaining stake
+    {
+      // allocation will be at 200% of max allowable capacity, so effective weight should now be 200%
+      const activeStake = await stakingPool.getActiveStake();
+      await burnStake.call(this, { amount: activeStake.div(2), start });
+    }
+
+    // recalculate effective weight
+    await stakingProducts.recalculateEffectiveWeights(this.poolId, DEFAULT_PRODUCTS);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.gt(
+      await stakingProducts.getTotalTargetWeight(this.poolId),
+    );
+    // 2000 * 2.00 = 4000
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(4000);
+  });
+
+  it('should fail to recalculate effective weight for product not in system', async function () {
+    const { stakingProducts, cover } = this;
+
+    await setStakedProducts.call(this, { productIds: [1, 2, 3] });
+    await expect(stakingProducts.recalculateEffectiveWeights(this.poolId, [9999999])).to.be.revertedWithCustomError(
```

Comment:
the 9999999 is fairly obvious here but i'd still use a named constant to highlight this is the `missingProductId`